text/0000-trim-std-os.md

```diff
@@ -0,0 +1,146 @@
+changes and don't immediately enable fixing bugs like using LFS on Linux:
+
+* Code such as `let a: ino_t = entry.ino()` would break as the `ino()` function
+  will return `u64`, but the definition of `ino_t` may not be `u64` for all
+  platforms.
+* The `stat` structure itself on 32-bit Linux still uses 32-bit fields (e.g. it
+  doesn't mirror `stat64` in libc).
+
+To help with these issues, more extensive modifications can be made to the
+platform specific modules. All type aliases can be switched over to `u64` and
+the `stat` structure could simply be redefined to `stat64` on Linux (minus
+keeping the same name). This would, however, explicitly mean that
+**std::os::raw is no longer FFI compatible with C**.
+
+This breakage can be clearly indicated in the deprecation messages, however.
+Additionally, this fits within std's [breaking changes policy][api-evolution] as
+a local `as` cast should be all that's needed to patch code that breaks to
+straddle versions of Rust.
+
+[api-evolution]: https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+As mentioned above, this RFC is strictly-speaking a breaking change. It is
+expected that not much code will break, but currently there is no data
+supporting this.
+
+Returning `u64` across the board could be confusing in some circumstances as it
+may wildly differ both in terms of signedness as well as size from the
+underlying C type. Converting it back to the appropriate type runs the risk of
+being onerous, but accessing these raw fields in theory happens quite rarely as
+std should primarily be exporting cross-platform accessors for the various
+fields here and there.
+
+# Alternatives
+[alternatives]: #alternatives
+
+* The documentation of the raw modules in std could be modified to indicate that
+  the types contained within are intentionally not FFI compatible, and the same
+  structure could be preserved today with the types all being rewritten to what
+  they would be anyway if this RFC were implemented. For example `ino_t` on
+  Android would change to `u64` and `stat` on 32-bit Linux would change to
+  `stat64`. In doing this, however, it's not clear why we'd keep around all the
+  C namings and structure.
+
+* Instead of breaking existing functionality, new accessors and types could be
+  added to acquire the "lossless" version of a type. For example we could add a
+  `ino64` function on `DirEntryExt` which returns a `u64`, and for `stat` we
+  could add `as_raw_stat64`. This would, however, force `Metadata` to store two
+  different `stat` structures, and the breakage in practice this will cause may
+  be small enough to not warrant these great lengths.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Is the policy of almost always returning `u64` too strict? Should types like
+  `mode_t` be allowed as `i32` explicitly? Should the sign at least attempt to
+  always be preserved?
```

Comment:
I'd aim for lossless casts only.  Widening should always be safe, but it would be bad to lose the sign on something like `time_t`.  For another example, `dev_t` is `u64` on Linux and `i32` on OSX -- there's no universal upcast for these (`i128`?), so it will have to have platform-specific signedness.  (Consider using integer `From`/`Into` instead of manual `as` casts to keep things safe.)